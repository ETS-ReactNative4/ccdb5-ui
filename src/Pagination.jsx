import React from 'react';

export default class Pagination extends React.Component {
  constructor(props) {
    super(props);
    this.state = this._calculatePages(props);

    // This binding is necessary to make `this` work in the callback
    // https://facebook.github.io/react/docs/handling-events.html
    this._handleSubmit = this._handleSubmit.bind(this);
    this._updateInputValue = this._updateInputValue.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    this.setState(this._calculatePages(nextProps));
  }

  _calculatePages(props) {
    const from = props.from || 0;
    const size = props.size || 1;
    const total = props.total || 0;
    const c = Math.ceil(from / size) + 1;

    return {
      current: c,
      total: Math.ceil(total / size),
      inputValue: c
    }
  }

  _handleSubmit(event) {
    event.preventDefault();
    this._setPage(this.state.inputValue);
  }

  _setPage(page) {
    this.props.onPage(page);
  }

  _updateInputValue(evt) {
    this.setState({
      inputValue: evt.target.value
    });
  }

  render() {
    if( this.props.total === 0) {
      return null;
    }

    return (
      <nav className="m-pagination"
           role="navigation"
           aria-label="Pagination">
          <button className="a-btn
                    a-btn__icon-on-left
                    cf-icon
                    cf-icon-left
                    cf-icon__before
                    m-pagination_btn-prev"
                    onClick={() => this._setPage(this.state.current - 1)}
                    disabled={this.state.current <= 1}>
              Previous
          </button>
          <button className="a-btn
                    a-btn__icon-on-right
                    cf-icon
                    cf-icon-right
                    cf-icon__after
                    m-pagination_btn-next"
                    onClick={() => this._setPage(this.state.current + 1)}
                    disabled={this.state.current >= this.state.total}>
              Next
          </button>
          <form className="m-pagination_form"
                action="" onSubmit={this._handleSubmit}>
              <label className="m-pagination_label"
                     htmlFor="m-pagination_current-page">
                  Page
                  <span className="u-visually-hidden">
                      number {this.state.current} out
                  </span>
                  <input className="m-pagination_current-page"
                         id="m-pagination_current-page"
                         name="page"
                         type="number"
                         min="1"
                         max={this.state.total}
                         pattern="[0-9]*"
                         inputMode="numeric"
                         value={this.state.inputValue}
                         onChange={this._updateInputValue} />
                  of {this.state.total}
              </label>
              <button className="a-btn
                             a-btn__link
                             m-pagination_btn-submit"
                      id="m-pagination_btn-submit"
                      type="submit">Go</button>
          </form>
      </nav>
    );
  }
}