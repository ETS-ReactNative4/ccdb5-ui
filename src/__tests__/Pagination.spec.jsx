import { IntlProvider } from 'react-intl'
import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import {shallow} from 'enzyme'
import { Pagination, mapStateToProps, mapDispatchToProps } from '../Pagination'

describe('component::Pagination', () => {
  it('renders without crashing', () => {
    const target = renderer.create(
      <IntlProvider locale="en">
        <Pagination page={1} total={100} />
      </IntlProvider>
    )
    let tree = target.toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('buttons', () => {
    let onCb = null
    let nextCb = null
    let prevCb = null
    let target = null

    beforeEach(() => {
      onCb = jest.fn()
      nextCb = jest.fn()
      prevCb = jest.fn()

      window.scrollTo = jest.fn();
      target = shallow(<Pagination onPage={onCb}
                                   nextPage={nextCb}
                                   prevPage={prevCb}/>)
    })

    it('is called when the previous button is clicked', () => {
      const prev = target.find('.m-pagination_btn-prev')
      prev.simulate('click')
      expect(onCb).not.toHaveBeenCalled()
      expect(nextCb).not.toHaveBeenCalled()
      expect(prevCb).toHaveBeenCalled()
    })

    it('is called when the next button is clicked', () => {
      const next = target.find('.m-pagination_btn-next')
      next.simulate('click')
      expect(onCb).not.toHaveBeenCalled()
      expect(prevCb).not.toHaveBeenCalled()
      expect(nextCb).toHaveBeenCalled()
    })

    it('is called when the form is submitted', () => {
      target.setState({
        page: 8
      });
      const theForm = target.find('form')
      theForm.simulate('submit', { preventDefault: () => {} })
      expect(prevCb).not.toHaveBeenCalled()
      expect(nextCb).not.toHaveBeenCalled()
      expect(onCb).toHaveBeenCalledWith(8)
    })
  })

  it('records text input from the user', () => {
    const target = shallow(<Pagination />)
    const textInput = target.find('[type="number"]')
    target.setState({
      page: 8
    });
    expect(target.state('page')).toEqual(8)
    // Note: out of range
    textInput.simulate('change', {target: { value: 49}})
    expect(target.state('page')).toEqual(49)
  })

  describe('button states', () => {
    it('enables the previous button when it is past first page', () => {
      const target = shallow(<Pagination page={3} total={10} />)
      const prev = target.find('.m-pagination_btn-prev')
      expect(prev.props().disabled).toEqual(false)
    })

    it('disables the previous button when on the first page', () => {
      const target = shallow(<Pagination page={1} total={10} />)
      const prev = target.find('.m-pagination_btn-prev')
      expect(prev.props().disabled).toEqual(true)
    })


    it('enables the next button when not on the last page', () => {
      const target = shallow(<Pagination page={2} total={10} />)
      const next = target.find('.m-pagination_btn-next')
      expect(next.props().disabled).toEqual(false)
    })

    it('disables the next button when on the last page', () => {
      const target = shallow(<Pagination page={10} total={10} />)
      const next = target.find('.m-pagination_btn-next')
      expect(next.props().disabled).toEqual(true)
    })
  })

  describe('mapDispatchToProps', () => {
    it('hooks into onPage', () => {
      const dispatch = jest.fn()
      mapDispatchToProps(dispatch).onPage(99)
      expect(dispatch.mock.calls.length).toEqual(1)
    })
    it('hooks into nextPage', () => {
      const dispatch = jest.fn()
      mapDispatchToProps(dispatch).nextPage()
      expect(dispatch.mock.calls.length).toEqual(1)
    })
    it('hooks into prevPage', () => {
      const dispatch = jest.fn()
      mapDispatchToProps(dispatch).prevPage()
      expect(dispatch.mock.calls.length).toEqual(1)
    })
  })

  describe('mapStateToProps', () => {
    it('maps state and props', () => {
      const state = {
        query: {
          page: 1,
          size: 25,
          totalPages: 100
        }
      }
      let actual = mapStateToProps( state )
      expect( actual ).toEqual( { page: 1, size: 25, total: 100 } )
    });
  })
})


