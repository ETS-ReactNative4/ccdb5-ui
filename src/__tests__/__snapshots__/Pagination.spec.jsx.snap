exports[`initial state renders without crashing 1`] = `
<nav
  aria-label="Pagination"
  className="m-pagination"
  role="navigation">
  <button
    className="a-btn a-btn__icon-on-left cf-icon cf-icon-left cf-icon__before m-pagination_btn-prev"
    disabled={false}
    onClick={[Function]}>
    Previous
  </button>
  <button
    className="a-btn a-btn__icon-on-right cf-icon cf-icon-right cf-icon__after m-pagination_btn-next"
    disabled={false}
    onClick={[Function]}>
    Next
  </button>
  <form
    action=""
    className="m-pagination_form"
    onSubmit={[Function]}>
    <label
      className="m-pagination_label"
      htmlFor="m-pagination_current-page">
      Page
      <span
        className="u-visually-hidden">
        number 
        8
         out
      </span>
      <input
        className="m-pagination_current-page"
        id="m-pagination_current-page"
        inputMode="numeric"
        max={40}
        min="1"
        name="page"
        onChange={[Function]}
        pattern="[0-9]*"
        type="number"
        value={8} />
      of 
      40
    </label>
    <button
      className="a-btn a-btn__link m-pagination_btn-submit"
      id="m-pagination_btn-submit"
      type="submit">
      Go
    </button>
  </form>
</nav>
`;
